libcob/intrinsic.c

```
@@ -4333,15 +4332,26 @@ cob_intr_char (cob_field *srcfield)
 	if (i < 1 || i > 256) {
 		*curr_field->data = 0;
 	} else {
-		*curr_field->data = (unsigned char)i - 1;
+		const unsigned char *col = COB_MODULE_PTR->collating_sequence;
```

What happens when the collating sequence doesn't have 255 elements?

Note: here and in the "chr == 0" because of loop case below we need to raise an argument exception:

> The value of argument-1 shall be greater than zero and less than or equal to the number of positions 
in the alphanumeric program collating sequence.

The old code should have read:

```c
	if (i < 1 || i > 256) {
		cob_set_exception (COB_EC_ARGUMENT_FUNCTION);
		*curr_field->data = 0;
	} else {
		cobglobptr->cob_exception_code = 0;
		*curr_field->data = (unsigned char)i - 1;
	}
```

please apply it to this change (adding it in the changelog)

---

> What happens when the collating sequence doesn't have 255 elements?

It always has. If specifying a custom alphabet with less than 255 characters, the remaining positions are filled with all the characters that were not explicitly specified.

> Note: here and in the "chr == 0" because of loop case below we need to raise an argument exception
> 
> > The value of argument-1 shall be greater than zero and less than or equal to the number of positions
> > in the alphanumeric program collating sequence.

Not sure about the `chr == 0` case ; it would imply the collating sequence is not a bijection, but we know it is by construction (and if it isn't it would probably be an internal error instead of an exception).

---

For the collation that makes sense, for the standard exception... we'd have to add the size.... maybe just add a new cob_intr_char_col that would take the data, collation and collation size (and ~~forward from the old to the new one in 3.x~~ would also remove the need to check for the collation in intr_char)?